compiler/shared/main/scala/mlscript/compiler/ClassLifter.scala

```diff
@@ -0,0 +1,482 @@
+
+import mlscript.*
+import mlscript.utils.StringOps
+import scala.collection.mutable.StringBuilder as StringBuilder
+import scala.collection.mutable.Map as MMap
+import scala.collection.mutable.Set as MSet
+import mlscript.codegen.Helpers.inspect as showStructure
+
+class ClassLifter { self: ClassLifter =>
+
+  type ClassName = String
+  type FieldName = String
+  case class LocalContext(vSet: Set[Var], tSet: Set[TypeName]){
+    def ++(rst: LocalContext) = LocalContext(vSet ++ rst.vSet, tSet ++ rst.tSet)
+    def -+(rst: LocalContext) = LocalContext(vSet -- rst.vSet, tSet ++ rst.tSet)
+    def addV(rst: IterableOnce[Var]) = LocalContext(vSet ++ rst, tSet)
+    def addV(nV: Var) = LocalContext(vSet + nV, tSet)
+    def extV(rst: IterableOnce[Var]) = LocalContext(vSet -- rst, tSet)
+    def extV(nV: Var) = LocalContext(vSet - nV, tSet)
+    def addT(rst: IterableOnce[TypeName]) = LocalContext(vSet, tSet ++ rst)
+    def addT(nT: TypeName) = LocalContext(vSet, tSet + nT)
+    def extT(rst: IterableOnce[TypeName]) = LocalContext(vSet, tSet -- rst)
+    def vSet2tSet = LocalContext(Set(), vSet.map(t => TypeName(t.name)))
+    def moveT2V(ts: Set[TypeName]) = {
+      val inters = tSet.intersect(ts)
+      LocalContext(vSet ++ inters.map(x => Var(x.name)), tSet -- inters)
+    }
+    def intersect(rst: LocalContext) = LocalContext(vSet intersect rst.vSet, tSet intersect rst.tSet)
+    override def toString(): String = "(" ++ vSet.mkString(", ") ++ "; " ++ tSet.mkString(", ") ++ ")"
+  }
+  private def asContext(v: Var) = LocalContext(Set(v), Set())
+  private def asContextV(vS: IterableOnce[Var]) = LocalContext(vS.toSet, Set())
+  private def asContext(t: TypeName) = LocalContext(Set(), Set(t))
+  private def asContextT(tS: IterableOnce[TypeName]) = LocalContext(Set(), tS.toSet)
+  private def emptyCtx = LocalContext(Set(), Set())
+
+  case class ClassInfoCache(
+      originNm: TypeName, 
+      liftedNm: TypeName, 
+      var capturedParams: LocalContext, 
+      fields: MSet[Var], 
+      innerClses: MMap[TypeName, ClassInfoCache], 
+      supClses: Set[TypeName],
+      outerCls: Option[ClassInfoCache],
+      body: NuTypeDef, 
+      depth: Int){
+        def ++(other: ClassInfoCache) = this.copy(
+            capturedParams = this.capturedParams ++ other.capturedParams, 
+            fields = this.fields ++ other.fields, 
+            innerClses = this.innerClses ++ other.innerClses
+          )
+        override def toString(): String = liftedNm.name ++ "@" ++ capturedParams.toString() ++ "^" ++ outerCls.map(_.liftedNm.toString()).getOrElse("_")
+      }
+  type ClassCache = Map[TypeName, ClassInfoCache]
+  type NamePath = List[String]
+
+  var retSeq: List[NuTypeDef] = Nil
+  var anonymCnt: Int = 0
+  var clsCnt: Int = 0
+  val logOutput: StringBuilder = new StringBuilder
+  val primiTypes = new mlscript.Typer(false, false, false).primitiveTypes
+
+  private def log(str: String): Unit = {
+    logOutput.append(str)
+    // println(str)
```

Comment:
This is only for helping with debugging…if needed, as the whole “log” method doesn’t serve for the main logic, it can be removed as well. Should I also do that?

Reply:
No, of course not. Having the ability of performing debugging is useful. Leftover comments however are not.